google-cloud-spanner/src/main/java/com/google/cloud/spanner/ThreadFactoryUtil.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.api.core.InternalApi;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import javax.annotation.Nullable;
+
+/** Utility class for creating a thread factory for daemon or virtual threads. */
+@InternalApi
+public class ThreadFactoryUtil {
+
+  /**
+   * Tries to create a thread factory for virtual threads, and otherwise falls back to creating a
+   * platform thread factory that creates daemon threads. Virtual threads are supported from JDK21.
+   *
+   * @param baseNameFormat the base name format for the threads, '-%d' will be appended to the
+   *     actual thread name format
+   * @param tryVirtualThreads whether to try to use virtual threads if available or not
+   * @return a {@link ThreadFactory} that produces virtual threads (Java 21 or higher) or daemon
+   *     threads
+   */
+  @InternalApi
+  public static ThreadFactory createVirtualOrDaemonThreadFactory(
+      String baseNameFormat, boolean tryVirtualThreads) {
+    ThreadFactory virtualThreadFactory =
+        tryVirtualThreads ? tryCreateVirtualThreadFactory(baseNameFormat) : null;
+    if (virtualThreadFactory != null) {
+      return virtualThreadFactory;
+    }
+
+    return new ThreadFactoryBuilder().setDaemon(true).setNameFormat(baseNameFormat + "-%d").build();
+  }
+
+  /**
+   * Tries to create a {@link ThreadFactory} that creates virtual threads. Returns null if virtual
+   * threads are not supported on this JVM.
+   */
+  @InternalApi
+  @Nullable
+  public static ThreadFactory tryCreateVirtualThreadFactory(String baseNameFormat) {
```

Comment:
Does this needs to be public?

Reply:
Yes, because we want to use it in a number of places:
1. In this package to create a virtual thread executor for gRPC (if configured). For that, it could have been package-private.
2. In the `.connection` package, so we can use it to create the thread that is used by connections to execute statements.
3. In some of the libraries that depend on this client library. Specifically in PGAdapter and the JDBC driver.